Don’t send email from callbacks (#48)
andycroll/andycroll.com: _posts/2019/2019-06-02-dont-email-from-activerecord-callbacks.md

---
title: "Don’t Email From Active Record Callbacks"
description: "It’s in the docs, but it’s a bad idea."
layout: article
category: ruby
image:
  base: '2019/dont-email-from-activerecord-callbacks'
  alt: 'Envelopes'
  credit: Joanna Kosinska
  source: "https://unsplash.com/photos/uGcDWKN91Fs"

---

One of the first things you often want to do within your Rails application is send email.

A frequent pattern is that an email is sent after a change to, or creation of, an instance of one of your models.

## Instead of…

> …sending email inside a callback in your model.

Author: ```suggestion
…sending email inside a callback in your model:
```

```ruby
class Comment < ApplicationRecord
  after_create :send_email_to_author

  private

  def send_email_to_author
    AuthorMailer.
      with(author: author).
      comment_notification.
      deliver_now
  end
end
```

## Use…

…the controller to send your email.

```ruby
class CommentsController < ApplicationController
  def create
    Comment.create(comment_params)
    AuthorMailer.
      with(author: author).
      comment_notification.
      deliver_now
  end
end
```

## Why?

This is all about clarity and preventing unintended side effects.

> At some point, you’ll likely need to create a comment without emailing the author; perhaps in the `rails console` or in another action. You _can_ use methods that [skip callbacks](https://guides.rubyonrails.org/active_record_callbacks.html#skipping-callbacks), but then you’re into a whole new world of complexity.

Reviewer: *Let's consider the example above. At some point,...

> It’s clearer to have a simple, procedural, 'list of things to do' inside controller actions, as you can read through from A to Z. I also find that debugging side-effects through a model’s callbacks to be much more cognitively difficult, as you have to keep more context—from more files—in your head.

Reviewer: ```suggestion
It’s clearer to have a simple, procedural, 'list of things to do' inside controller actions that you can read through from A to Z. Also, debugging side-effects through a model’s callbacks is much more cognitively difficult, as you have to keep more context—from more files—in your head.
```

## Why not?

> Using callbacks in models to send email could be considered “the Rails Way”, given examples like this have always been part of the documentation. There’s no _real_ harm in this simple case, but the pain of this approach often only becomes apparent later.

Reviewer: *no comma before 'given'

Reviewer: *often becomes apparent later,  once you have more complex scenarios (?) — or something like that.

> There’s a frequently stated opinion preferring “fat models & thin controllers”; pushing as much functionality as you can into your model layer.

Reviewer: This paragraph and the next should be one:

"There’s a frequently-stated opinion that says prefer “fat models and thin controllers”, pushing as much functionality as you can into your model layer. The focus of this saying is more about keeping complexity out of the controller layer, rather than advocating for more use of callbacks in models, which create side-effects. I’d also argue that sending an email within a controller, in addition to the model change, isn’t a complex series of operations."

Also — what exactly are the 'side effects' that you keep referring to? I'm not sure I'm clear.

---
title: "Don’t Email From Active Record Callbacks"
description: "It’s in the docs, but it’s a bad idea."
layout: article
category: ruby
image:
  base: '2019/dont-email-from-activerecord-callbacks'
  alt: 'Envelopes'
  credit: Joanna Kosinska
  source: "https://unsplash.com/photos/uGcDWKN91Fs"

---

One of the first things you often want to do within your Rails application is send email.

A frequent pattern is that an email is sent after a change to, or creation of, an instance of one of your models.

## Instead of…

…sending email inside a callback in your model:

```ruby
class BookReview < ApplicationRecord
  after_create :send_email_to_author

  private

  def send_email_to_author
    AuthorMailer.
      with(author: author).
      review_notification.
      deliver_now
  end
end
```

## Use…

…the controller to send your email.

```ruby
class BookReviewsController < ApplicationController
  def create
    BookReview.create(comment_params)
    AuthorMailer.
      with(author: author).
      review_notification.
      deliver_now
  end
end
```

## Why?

This is all about clarity and preventing surprises when you come back to the code.

> Let's consider the example above. Creating a review doesn’t _always_ need to send an email to author of the book. In the first example, the email sending happens as a side effect of creating a new book review.

Reviewer: *to the author

Reviewer: *In the first code sample..

> At some point, you’ll likely need to create one without emailing the author; perhaps in the `rails console` or in another action. You _can_ use methods that [skip callbacks](https://guides.rubyonrails.org/active_record_callbacks.html#skipping-callbacks), but then you’re into a whole new world of complexity.

Reviewer: *create a new review (if it's a separate paragraph)

It’s clearer to have a simple, procedural, 'list of things to do' inside controller actions that you can read through from A to Z. In this case, having the email sending be a separate _thing_ that happens after the creation of the review makes things much more obvious when you come back to the code later.

In addition, debugging tangential functionality through a model’s callbacks is much more cognitively difficult, as you have to keep more context—from more files—in your head.

## Why not?

Using callbacks in models to send email could be considered “the Rails Way” given examples like this have always been part of the documentation. And there’s no _real_ harm in this simple case, but the pain of this approach often only becomes apparent later as the complexity of your application increases.

> There’s a frequently stated colloquial recommendation that we should prefer “fat models & thin controllers”, pushing as much functionality as you can into your model layer. This is generally good advice. But, it is more about keeping complexity out of the controller layer, by making the activities of your application clear, rather than advocating for more use of side effect generating callbacks.

Reviewer: *_and_ think controllers (since it's a phrase, no..? Or is it often written with the ampersand? If so, keep it.)

Reviewer: *side effect-generating (I think...)

> I’d also argue that sending an email alongside a model change, is not complex enough to merit this confusing callback-based abstraction. It is important enough, to users of your application, to clearly surface that both things will happen when this controller action is called.

Reviewer: No comma after 'model change'.

Reviewer: To users or to other developers?

Reviewer: *to merit _a_ confusing callback-based...

Reviewer: *It is important, to the...., to clearly surface (so basically take out the 'enough').

Reviewer: *when _the_ controller action is called
